File: client/src/api/agents.ts
/**
 * AI Agent API client, status, config generation, PII detection,
 * rule explanation (SSE), and compliance analysis.
 */

import { fetchApi, getAuthHeaders } from "./client";

export interface PiiEnforcementStatus {
  configured: boolean;
  model: string;
  api_base: string;
  require_local: boolean;
  local_verified: boolean;
  detail: string;
}

export interface AgentStatus {
  enabled: boolean;
  provider: string;
  model: string;
  api_base: string;
  circuit_breaker: Record<string, unknown>;
  cache_size: number;
  pii_enforcement?: PiiEnforcementStatus;
}

export interface ConfigGenRequest {
  prompt: string;
  regulation?: string;
}

export interface ConfigGenResponse {
  yaml: string;
  valid: boolean;
  validation_error: string;
  source: string;
}

export interface PiiDetection {
  resource_id: string;
  resource_type: string;
  field_path: string;
  type: string;
  evidence: string;
  confidence: number;
  severity: string;
  source: string;
}

export interface PiiDetectionResponse {
  detections: PiiDetection[];
  summary: { total: number; critical: number; high: number; medium: number };
  layers_used: string[];
}

export interface ComplianceResponse {
  regulation: string;
  compliance_score: number;
  gaps: Array<{
    requirement: string;
    description: string;
    severity: string;
  }>;
  excess: Array<{ rule_match: string; reason: string }>;
  recommendations: Array<{
    priority: number;
    action: string;
    rule: Record<string, unknown>;
    reason: string;
  }>;
  source?: string;
}

export async function getAgentStatus(): Promise<AgentStatus> {
  return fetchApi<AgentStatus>("/v1/ai/status");
}

export async function generateConfig(
  req: ConfigGenRequest,
): Promise<ConfigGenResponse> {
  return fetchApi<ConfigGenResponse>("/v1/ai/generate-config", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(req),
  });
}

export async function detectPii(
  resources: Record<string, unknown>[],
  useAi: boolean = true,
  /** Shortest string value to scan. Default 15 targets free-text; pass 1 to
   * scan every string field (short SSN/phone/name in structured fields). */
  minFieldLen: number = 15,
): Promise<PiiDetectionResponse> {
  return fetchApi<PiiDetectionResponse>("/v1/ai/detect-pii", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ resources, use_ai: useAi, min_field_len: minFieldLen }),
  });
}

export async function explainConfig(
  yamlText: string,
  regulation?: string,
): Promise<{ explanation: string }> {
  return fetchApi<{ explanation: string }>("/v1/ai/explain", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ yaml_text: yamlText, regulation }),
  });
}

export async function* streamExplain(
  yamlText: string,
  regulation?: string,
): AsyncGenerator<string, void, undefined> {
  const response = await fetch("/api/v1/ai/explain", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "text/event-stream",
      ...getAuthHeaders(),
    },
    body: JSON.stringify({ yaml_text: yamlText, regulation }),
  });
  const reader = response.body?.getReader();
  if (!reader) return;
  const decoder = new TextDecoder();
  let buffer = "";
  for (;;) {
    const { done, value } = await reader.read();
    if (value) buffer += decoder.decode(value, { stream: true });
    let idx: number;
    while ((idx = buffer.indexOf("\n\n")) !== -1) {
      const line = buffer.slice(0, idx).trim();
      buffer = buffer.slice(idx + 2);
      if (line.startsWith("data: ")) {
        const data = line.slice(6);
        if (data === "[DONE]") return;
        try {
          const parsed = JSON.parse(data);
          if (parsed.text) yield parsed.text;
        } catch {
          /* skip malformed SSE chunks */
        }
      }
    }
    if (done) break;
  }
}

export interface ChatTurn {
  role: "user" | "assistant";
  content: string;
}

/**
 * How structured (object) PII fields are treated when proposing rules.
 * - "values" (default): one rule per identifying leaf sub-field
 *   (Patient.name.family), keeps the FHIR skeleton, blanks only the values.
 * - "whole": one rule on the parent path (Patient.name), removes the element.
 */
export type FieldGranularity = "values" | "whole";

export interface ChatRequest {
  question: string;
  config_yaml?: string;
  history?: ChatTurn[];
  model?: string;
  /** Field-path tree from uploaded examples or server samples. Paths + types
   * only by default; carries sample values (PHI) when `include_values` is set. */
  field_context?: string;
  /** When true, `field_context` includes sample values, the backend then
   * treats the call as a PHI payload and refuses any non-local AI endpoint. */
  include_values?: boolean;
  /** Leaf-vs-whole treatment of structured PII fields. */
  granularity?: FieldGranularity;
}

/**
 * Stream a config-chat answer as SSE text chunks. Throws on a server-sent
 * `error` event so the caller can surface it.
 */
export async function* streamChat(
  req: ChatRequest,
): AsyncGenerator<string, void, undefined> {
  const response = await fetch("/api/v1/ai/chat", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "text/event-stream",
      ...getAuthHeaders(),
    },
    body: JSON.stringify(req),
  });
  if (!response.ok) {
    throw new Error(`Chat request failed (HTTP ${response.status})`);
  }
  const reader = response.body?.getReader();
  if (!reader) return;
  const decoder = new TextDecoder();
  let buffer = "";
  for (;;) {
    const { done, value } = await reader.read();
    if (value) buffer += decoder.decode(value, { stream: true });
    let idx: number;
    while ((idx = buffer.indexOf("\n\n")) !== -1) {
      const line = buffer.slice(0, idx).trim();
      buffer = buffer.slice(idx + 2);
      // Comment lines (": keep-alive") are ignored.
      if (line.startsWith("data: ")) {
        const data = line.slice(6);
        if (data === "[DONE]") return;
        try {
          const parsed = JSON.parse(data);
          if (parsed.error) throw new Error(parsed.error);
          if (parsed.text) yield parsed.text;
        } catch (e) {
          if (e instanceof Error && e.message) throw e;
          /* skip malformed SSE chunks */
        }
      }
    }
    if (done) break;
  }
}

export interface PiiScanResult {
  path: string;
  is_pii: boolean;
  reason: string;
  suggested_action: string;
}

interface FieldScanResponse {
  results: PiiScanResult[];
  source: "ai" | "error";
  detail: string;
}

/**
 * Ask the AI to classify every field path in `fieldContext` as PII or not, and
 * suggest an action for each PII field. Uses the dedicated /v1/ai/scan-fields
 * endpoint which returns validated, structured JSON (no fragile prose parsing).
 *
 * Throws when the backend reports `source: "error"` (AI disabled/unreachable)
 * so the caller can surface the `detail` message.
 */
export async function scanFieldsForPii(
  fieldContext: string,
  opts: {
    model?: string;
    granularity?: FieldGranularity;
    includeValues?: boolean;
    guidance?: string;
  } = {},
): Promise<PiiScanResult[]> {
  const res = await fetchApi<FieldScanResponse>("/v1/ai/scan-fields", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      field_context: fieldContext,
      model: opts.model ?? "",
      granularity: opts.granularity ?? "values",
      include_values: opts.includeValues ?? false,
      guidance: opts.guidance ?? "",
    }),
  });
  if (res.source === "error") {
    throw new Error(res.detail || "PII scan failed");
  }
  return res.results;
}

export interface FieldSketchResponse {
  sketch: string;
  types: string[];
  leaf_count: number;
  included_count: number;
  truncated: boolean;
  source: string; // "fhir" | "inline" | "error"
  detail: string;
}

/**
 * Build a compact, PHI-safe schema sketch of the selected resource types: one
 * line per distinct leaf path with its type, presence frequency, cardinality,
 * and a value digest (shape/enum in the default PHI-free mode; raw samples only
 * when `includeValues` is set). The result is a richer `field_context` string
 * to pass to {@link streamChat} / {@link scanFieldsForPii}.
 *
 * Supply `resources` to sketch uploaded examples, or `resourceTypes` to have the
 * server sample the configured source FHIR server.
 */
export async function buildFieldSketch(opts: {
  resourceTypes?: string[];
  resources?: Record<string, unknown>[];
  nPerType?: number;
  includeValues?: boolean;
}): Promise<FieldSketchResponse> {
  return fetchApi<FieldSketchResponse>("/v1/ai/field-sketch", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      resource_types: opts.resourceTypes ?? [],
      resources: opts.resources ?? [],
      n_per_type: opts.nPerType ?? 25,
      include_values: opts.includeValues ?? false,
    }),
  });
}

export async function analyseCompliance(
  yamlText: string,
  regulation: string,
): Promise<ComplianceResponse> {
  return fetchApi<ComplianceResponse>("/v1/ai/compliance", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ yaml_text: yamlText, regulation }),
  });
}

File: services/anonymizer/src/api/schemas/agents.py
"""Request/response schemas for AI agent endpoints."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class AgentStatusResponse(BaseModel):
    enabled: bool
    provider: str = ""
    model: str = ""
    api_base: str = ""
    circuit_breaker: dict = Field(default_factory=dict)
    cache_size: int = 0
    pii_enforcement: dict = Field(default_factory=dict)


class ConfigGenerationRequest(BaseModel):
    prompt: str = Field(
        ...,
        min_length=10,
        max_length=2000,
        description="Natural language description of de-identification requirements.",
    )
    regulation: str = Field(
        default="",
        description="Target compliance framework (e.g., 'HIPAA', 'GDPR').",
    )
    include_source_context: bool = Field(
        default=False,
        description="When true, query the configured FHIR_SOURCE_URL for the "
        "resource types/counts present and ground the generated rules in that "
        "actual dataset (no patient data is read).",
    )


class ConfigGenerationResponse(BaseModel):
    model_config = {"extra": "allow"}

    yaml: str
    valid: bool
    validation_error: str = ""
    source: str  # "ai" | "fallback" | "error"


class PiiDetectionRequest(BaseModel):
    resources: list[dict] = Field(
        ...,
        min_length=1,
        max_length=100,
        description="De-identified FHIR resources to scan for residual PII.",
    )
    use_ai: bool = Field(
        default=True,
        description="Enable AI-powered contextual analysis (requires local model).",
    )
    min_field_len: int = Field(
        default=15,
        ge=1,
        le=1000,
        description=(
            "Shortest string value to scan. Default 15 targets free-text "
            "narrative; set 1 to scan every string field (catches short "
            "structured identifiers like SSN/phone/name at the cost of more "
            "NER noise)."
        ),
    )


class PiiDetectionResponse(BaseModel):
    model_config = {"extra": "allow"}

    detections: list[dict]
    summary: dict
    layers_used: list[str]


class ExplainRequest(BaseModel):
    yaml_text: str = Field(
        ...,
        min_length=10,
        description="YAML config to explain.",
    )
    regulation: str = Field(
        default="",
        description="Optional regulation for alignment analysis.",
    )


class ComplianceRequest(BaseModel):
    yaml_text: str = Field(
        ...,
        min_length=10,
        description="YAML config to analyse.",
    )
    regulation: str = Field(
        ...,
        min_length=2,
        description="Target regulation (e.g., 'HIPAA Safe Harbor', 'GDPR').",
    )


class FieldScanRequest(BaseModel):
    field_context: str = Field(
        ...,
        min_length=1,
        max_length=60000,
        description="Field-path tree to classify. Paths + JSON value types only "
        "by default; when ``include_values`` is true, each leaf also carries a "
        "truncated SAMPLE value (``path : <type> = value``). Server-derived → "
        "treated as untrusted DATA (sanitized + tag-wrapped before injection).",
    )
    include_values: bool = Field(
        default=False,
        description="When true, ``field_context`` carries patient sample values "
        "so a LOCAL model can judge PII more accurately. The call is then a PHI "
        "payload: the AI local-guard refuses any non-local endpoint (fail-"
        "closed), so values never leave a self-hosted model.",
    )
    model: str = Field(
        default="",
        description="Optional local model override (e.g. 'ollama/gemma3:1b').",
    )
    guidance: str = Field(
        default="",
        max_length=2000,
        description="Optional free-text user guidance on how to treat fields "
        "(e.g. 'pseudonymize all identifiers', 'generalize dates to year'). "
        "Injected as a sanitized system instruction the model follows when "
        "classifying  never mixed into the untrusted field tree.",
    )
    granularity: Literal["values", "whole"] = Field(
        default="values",
        description="How structured (object) fields are treated. 'values' "
        "(default) classifies leaf sub-fields (Patient.name.family) so the FHIR "
        "skeleton is kept and only values are blanked; 'whole' classifies the "
        "parent container (Patient.name) as a single row.",
    )


class FieldScanResult(BaseModel):
    path: str
    is_pii: bool
    reason: str = ""
    suggested_action: str = ""


class FieldScanResponse(BaseModel):
    results: list[FieldScanResult]
    source: str  # "ai" | "error"
    detail: str = ""


class FieldSketchRequest(BaseModel):
    resource_types: list[str] = Field(
        default_factory=list,
        max_length=64,
        description="Resource types to sample from the configured source FHIR "
        "server (e.g. ['Patient', 'Observation']). Ignored when ``resources`` "
        "is supplied.",
    )
    resources: list[dict] = Field(
        default_factory=list,
        max_length=1000,
        description="Pre-sampled FHIR resources (e.g. the user's uploaded "
        "examples). When present, the server sketches THESE instead of querying "
        "the source server; grouped by ``resourceType`` server-side.",
    )
    n_per_type: int = Field(
        default=25,
        ge=1,
        le=200,
        description="Max resources to sample per type when sampling the source "
        "server. Higher values catch rarer optional fields at more I/O cost.",
    )
    include_values: bool = Field(
        default=False,
        description="When true, each leaf carries a truncated raw SAMPLE value "
        "instead of a PHI-free shape/enum. The returned sketch then contains "
        "PHI, so it is not cached and, when later sent to a model, engages the "
        "local-guard (fail-closed).",
    )


class FieldSketchResponse(BaseModel):
    sketch: str = ""
    types: list[str] = Field(default_factory=list)
    leaf_count: int = 0
    included_count: int = 0
    truncated: bool = False
    source: str = ""  # "fhir" | "inline" | "error"
    detail: str = ""


class ChatMessage(BaseModel):
    role: str = Field(..., description="'user' or 'assistant'.")
    content: str = Field(..., max_length=8000)


class ChatIntake(BaseModel):
    """Structured intake collected by the UI before the first chat turn.

    Lets the assistant ground its proposals in the user's stated scope
    (resource types), compliance target, and free-text requirements WITHOUT
    smuggling that metadata into the user's question text. All fields are
    optional; an empty intake injects no extra context.
    """

    resource_types: list[str] = Field(
        default_factory=list,
        max_length=64,
        description="FHIR resource types the user wants rules for "
        "(e.g. ['Patient', 'Observation']). Empty = all types.",
    )
    regulation: str = Field(
        default="",
        max_length=64,
        description="Compliance standard the config must satisfy "
        "(e.g. 'HIPAA', 'GDPR', 'RESEARCH'). Empty/'CUSTOM' = no standard.",
    )
    intent: str = Field(
        default="",
        max_length=1000,
        description="Free-text requirements (e.g. 'keep dates as year-only, "
        "use gPAS for IDs').",
    )


class ChatRequest(BaseModel):
    question: str = Field(
        ...,
        min_length=1,
        max_length=2000,
        description="The user's question about the de-identification config.",
    )
    intake: ChatIntake | None = Field(
        default=None,
        description="Optional structured requirements (resource scope, "
        "regulation, free-text intent) gathered by the UI's guided intake "
        "step. Injected as dedicated system context, not mixed into the "
        "question text.",
    )
    config_yaml: str = Field(
        default="",
        max_length=60000,
        description="Optional current config YAML for context. Cap matches "
        "``field_context`` (the user's own authored config, already bounded by "
        "the request body cap); a comprehensive field-complete profile easily "
        "exceeds 20k chars. The prompt injects only a truncated prefix "
        "(``MEDANON_AI_CONFIG_CONTEXT_MAX``) to fit the model context.",
    )
    history: list[ChatMessage] = Field(
        default_factory=list,
        description="Prior turns for multi-turn context (most recent last).",
        max_length=20,
    )
    model: str = Field(
        default="",
        description="Optional model override (e.g. 'ollama/gemma3:1b'). "
        "Empty uses the configured default.",
    )
    include_source_context: bool = Field(
        default=False,
        description="When true, inject a PHI-free summary of the source FHIR "
        "server's resource types/counts so the assistant can tailor answers to "
        "the user's actual data.",
    )
    field_context: str = Field(
        default="",
        max_length=60000,
        description="Field-path tree extracted client-side from uploaded "
        "examples OR sampled from the live FHIR server, so the assistant targets "
        "proposed rules at the FHIRPaths the user actually has. Paths + types "
        "only by default; when ``include_values`` is true each leaf also carries "
        "a truncated SAMPLE value. Server-derived, so treated as untrusted DATA "
        "(sanitized + tag-wrapped before prompt injection).",
    )
    include_values: bool = Field(
        default=False,
        description="When true, ``field_context`` carries patient sample values "
        "so a LOCAL model can ground answers in real data. The call is then a "
        "PHI payload: the AI local-guard refuses any non-local endpoint (fail-"
        "closed), so values never leave a self-hosted model.",
    )
    granularity: Literal["values", "whole"] = Field(
        default="values",
        description="How structured (object) PII fields are treated when "
        "proposing rules. 'values' (default) emits one rule per identifying "
        "leaf sub-field (Patient.name.family) so the FHIR skeleton is preserved "
        "and only values are blanked; 'whole' emits one rule on the parent path "
        "(Patient.name) that removes the entire element.",
    )
